library/aesni.h

```diff
@@ -67,8 +73,13 @@
  * In the long run, we will likely remove the assembly implementation. */
 #if defined(MBEDTLS_AESNI_HAVE_INTRINSICS)
 #define MBEDTLS_AESNI_HAVE_CODE 2 // via intrinsics
-#elif defined(MBEDTLS_HAVE_X86_64)
+#elif defined(MBEDTLS_HAVE_ASM) && \
+    defined(__GNUC__) && defined(MBEDTLS_HAVE_X86_64)
 #define MBEDTLS_AESNI_HAVE_CODE 1 // via assembly
+#elif defined(__GNUC__)
+#   error "Must use `-mpclmul -msse2 -maes` for MBEDTLS_AESNI_C"
```

Comment:
I prefer add `#pragma GCC target("pclmul,sse2,aes")` in next PR to keep consistent with `aesce.c`,`sha*.c`. This should check compiler compatible, I do not think we should do that in this PR.

These checks should be moved to C file like other modules.  And I do not think the module should be disabled silently.

Reply:
We were planning to _maybe_ remove the assembly code. It's still useful for older runtimes that don't have the intrinsics. The question is, do we still care about these old runtimes?

Reply:
Also the assembly still gets used when compiling without the `-m` flags that enable them.

Reply:
gcc 4.6.1 is from 2011, 12 years ago. It is reasonable for development to require a compiler from the last 10 years.

We certainly don't need to go back that far for clang - in 2014 clang 3.5 couldn't build 95% of the Debian archive, and it wasn't until 2018 that it was used to build Firefox for Windows. I have a Docker environment for clang 6.0.1-14 on x86_64, which seems a reasonable oldest to support, although I would be happy with requiring an even more recent minimum version if necessary

Reply:
Could you share the environment?